### PR TITLE
emit container status events after network reconnection (fixes #13524)

### LINKED ISSUE #13524: [BUG] misleading status for non-host-mode containers
### Description

I've noticed something odd but I don't know whether it's actually a bug or the intended behaviour.

Consider this `diff`:

```
$ diff docker-compose.override.yml.save docker-compose.override.yml
4a5,14
>   default:
>     ipam:
>       config:
>         - subnet: 172.30.0.0/22
> 
>   nextcloud:
>     ipam:
>       config:
>         - subnet: 172.30.4.0/22
> 
```

In other words, `docker-compose.override.yml` defines specific subnets for two internal networks whereas the `.save` file does not.

The situation with those subnet definitions in force:

```
$ docker ps
CONTAINER ID   IMAGE                       COMMAND                  CREATED         STATUS                   PORTS                    NAMES
a3d379eb6ba7   koenkk/zigbee2mqtt:latest   "docker-entrypoint.s…"   2 minutes ago   Up 2 minutes             0.0.0.0:8080->8080/tcp   zigbee2mqtt
c4e47e0f236d   iotstack-nodered            "./entrypoint.sh"        2 minutes ago   Up 2 minutes (healthy)   0.0.0.0:1880->1880/tcp   nodered
c484a9c1b5e9   grafana/grafana             "/run.sh"                2 minutes ago   Up 2 minutes (healthy)   0.0.0.0:3000->3000/tcp   grafana
ff51c96674a3   esphome/esphome             "/entrypoint.sh dash…"   2 minutes ago   Up 2 minutes (healthy)                            esphome
f127b0a10b85   iotstack-mosquitto          "/docker-entrypoint.…"   2 minutes ago   Up 2 minutes (healthy)   0.0.0.0:1883->1883/tcp   mosquitto
d8d58efc1c9d   influxdb:1.12               "/entrypoint.sh infl…"   2 minutes ago   Up 2 minutes (healthy)   0.0.0.0:8086->8086/tcp   influxdb
```

Now let's change the override file to be the one **without** the subnet definitions:

```
$ mv docker-compose.override.yml.save docker-compose.override.yml
```

Let's apply that:

```
$ docker compose up -d
WARN[0000] No services to build                         
[+] up 7/7
 ✔ Container nodered        Stopped 0.7s 
 ✔ Container zigbee2mqtt    Stopped 9.6s 
 ✔ Container grafana        Stopped 0.7s 
 ✔ Container influxdb       Stopped 1.0s 
 ✔ Container mosquitto      Stopped 0.4s 
 ✔ Network iotstack_default Created 0.1s 
 ✔ Container esphome        Running 0.0s
```

Compose says it has stopped five containers. The true situation:

``` 
$ docker ps
CONTAINER ID   IMAGE                       COMMAND                  CREATED         STATUS                             PORTS                    NAMES
a3d379eb6ba7   koenkk/zigbee2mqtt:latest   "docker-entrypoint.s…"   3 minutes ago   Up 12 seconds                      0.0.0.0:8080->8080/tcp   zigbee2mqtt
c4e47e0f236d   iotstack-nodered            "./entrypoint.sh"        3 minutes ago   Up 12 seconds (health: starting)   0.0.0.0:1880->1880/tcp   nodered
c484a9c1b5e9   grafana/grafana             "/run.sh"                3 minutes ago   Up 12 seconds (healthy)            0.0.0.0:3000->3000/tcp   grafana
ff51c96674a3   esphome/esphome             "/entrypoint.sh dash…"   3 minutes ago   Up 3 minutes (healthy)                                      esphome
f127b0a10b85   iotstack-mosquitto          "/docker-entrypoint.…"   3 minutes ago   Up 12 seconds (healthy)            0.0.0.0:1883->1883/tcp   mosquitto
d8d58efc1c9d   influxdb:1.12               "/entrypoint.sh infl…"   3 minutes ago   Up 12 seconds (health: starting)   0.0.0.0:8086->8086/tcp   influxdb
```

To summarise:

1. We started with 6 containers running (esphome, grafana, influxdb, mosquitto, nodered and zigbee2mqtt).
2. We finished with the same six containers running.
3. After changing the override file, all bar esphome are reported as having been stopped. esphome being the exception is expected because that container runs in host mode so it isn't affected by internal network changes.
4. What's odd is that docker compose hasn't reported **starting** the stopped containers again, yet they're all clearly running.

Compare and contrast what happens if I pick a container at random (eg grafana) and add some dummy environment variable to its service definition. On the subsequent "up", I'll get:

```
 ✔ Container grafana     Recreated 0.5s 
```

In short, the expected behaviour is that all five non-host-mode containers would be reported as **either** "Recreated" **or** "Stopped" then "Created", rather than just "Stopped" which isn't actually true.

For the record, this pattern is the same in each direction. Going from "no subnets defined" to "subnets defined" has the same behaviour. It also doesn't matter whether the subnet definition occurs in the override file or the compose file. I just happen to be using the override file.

I hope that makes sense.

### Steps To Reproduce

Implied in the above. In a compose file with a `networks:` section plus at least one non-host-mode container:

* either:

	1. Compose or override file **without** defined subnets.
	2. Down stack
	3. Up stack.
	3. Add subnet definitions to compose or override file.
	4. Up the stack.

* or:

	1. Compose or override file **with** defined subnets.
	2. Down stack
	3. Up stack.
	3. Remove subnet definitions from compose or override file.
	4. Up the stack.

### Compose Version

```Text
$ docker compose version
Docker Compose version v5.0.1

$ docker-compose version
Docker Compose version v5.0.1
```

### Docker Environment

```Text
Client: Docker Engine - Community
 Version:    29.1.3
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.30.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v5.0.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 6
  Running: 6
  Paused: 0
  Stopped: 0
 Images: 7
 Server Version: 29.1.3
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: local
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 CDI spec directories:
  /etc/cdi
  /var/run/cdi
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: dea7da592f5d1d2b7755e3a161be07f43fad8f75
 runc version: v1.3.4-0-gd6d73eb8
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.21-v8+
 Operating System: Debian GNU/Linux 12 (bookworm)
 OSType: linux
 Architecture: aarch64
 CPUs: 4
 Total Memory: 3.704GiB
 Name: sec-dev
 ID: f6406dbb-2fc1-4834-9606-5e826871aec6
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
  ::1/128
 Live Restore Enabled: false
 Firewall Backend: iptables
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by MaheshThakur9152**
Summary
- When a network configuration (for example IPAM/subnet) changes, Compose stops containers and recreates the network.
- The Docker engine can restart those containers automatically (e.g., due to a restart policy). Compose previously did not emit any event representing the containers' resumed state after reconnect, so the TTY UI could be left showing only "Stopped" even though the containers were actually running.
- This PR makes Compose inspect containers after reconnecting them and emits the correct container status (Running / Created / Stopped), so the UI reflects the real state.

Root cause
- The code path that handles diverged networks stops and disconnects containers, removes the network, recreates it and then reconnects containers.
- If the engine restarts those containers between disconnection and reconnection, Compose had no further event for them, so the TTY progress UI only displayed the earlier `Stopped`.

What I changed
- After reconnecting previously-dangled containers, inspect each container and emit one of the following events depending on the container state:
  - Running -> `Running` event
  - Created -> `Created` event
  - Exited -> `Stopped` event
- Ignore inspect errors to avoid failing network creation due to races where a container disappears between disconnect and reconnect.
- Add unit test `TestResolveOrCreateNetworkEmitsRunningEvent` to assert that a running event is emitted for a container restarted by the engine.

Files changed
- create.go — emits container status events after reconnecting containers
- create_test.go — new unit test simulating the diverged network flow and the engine restarting a container

Testing done
- Unit test: go test ./pkg/compose -run TestResolveOrCreateNetworkEmitsRunningEvent -v — passes locally.
- Full test suite: will run in CI (needs Docker daemon access and runs in their environments).

Notes for reviewers
- Small, focused change to event emission — no new event types or UI changes.
- Emitted events use existing statuses (Running, Created, Stopped) so the current UI picks them up.
- If maintainers prefer a different message (e.g., “Reconnected”), I’m happy to adjust.
- Signed-off-by: Mahesh Thakur <maheshthakur9152@gmail.com>

Fixes: docker/compose#13524

### CODE REVIEW

File: pkg/compose/create.go [hunk 1]
```diff
@@ -1420,6 +1420,23 @@ func (s *composeService) resolveOrCreateNetwork(ctx context.Context, project *ty
...
 	}
 
+	// Emit container status events for containers that may have been restarted by the engine
+	for _, c := range dangledContainers {
+		inspected, err := s.apiClient().ContainerInspect(ctx, c.ID)
+		if err != nil {
+			// ignore inspect errors
+			continue
+		}
+		name := "Container " + strings.TrimPrefix(inspected.Name, "/")
```
Comment (ndeloof) [inline pkg/compose/create.go]: use getCanonicalContainerName for consistency with other places we produce events

File: pkg/compose/create_test.go [hunk 2]
```diff
@@ -455,3 +455,75 @@ volumes:
 		})
 	}
 }
+
+func TestResolveOrCreateNetworkEmitsRunningEvent(t *testing.T) {
```
Comment (ndeloof) [inline pkg/compose/create_test.go]: We prefer use of e2e tests as mocks are a pain to maintain with many API calls, also in most case we don't get the actual engine behavior demonstrated by such tests.

Comment (MaheshThakur9152) [reply pkg/compose/create_test.go]: Hi — thanks for the review and the suggestions!

I implemented both fixes you pointed out:
- Switched to using the canonical container name (now using `getContainerProgressName(tmp)`) so emitted event names match other code paths.
- Fixed the state check to consistently use the `Status` field (`inspected.State.Status == container.StateRunning`), consistent with other code.

Per your preference for e2e tests, I added an e2e test:
- `TestNetworkIPAMChangeReportsStarted` changes the subnet and runs `docker compose up --progress=plain`, then asserts the output contains the expected container state transition (Stopped then Started or Recreated).

I left the focused unit test as an additional guard; if you prefer to rely only on e2e tests I can remove the unit test in a follow-up. Happy to tweak wording or add more coverage if you prefer.

— Mahesh

**Comment by ndeloof** [comment]
docker engine to automatically restart container is a corner case. The actual issue on https://github.com/docker/compose/issues/13524 is Compose not to log an event after a container is (re)started.

**Comment by MaheshThakur9152** [comment]
> docker engine to automatically restart container is a corner case. The actual issue on #13524 is Compose not to log an event after a container is (re)started.

Ack, understood. The auto-restart is just the trigger, but the missing event log is the gap in Compose. 

This PR addresses that root cause: it inspects the container state after network reconciliation and emits the missing status event.

I've updated the logic and added the E2E test to cover this. Ready for review.

**Comment by ndeloof** [comment]
> This PR addresses that root cause

it doesn't. It just emits an additional event, while this one should be managed by `start`

**Comment by MaheshThakur9152** [comment]
> > This PR addresses that root cause
> 
> it doesn't. It just emits an additional event, while this one should be managed by `start`

Got it. So instead of manually emitting the event post-reconnect, we should ensure the standard start lifecycle handles this transition?

Could you point me to the specific start method/flow you are referring to? I want to make sure I hook into the correct lifecycle stage rather than patching it here.

**Comment by ndeloof** [comment]
this is intentional: https://github.com/docker/compose/blob/c8d687599a425502116dac1de9c55d8ecc52bb77/cmd/display/tty.go#L193

if we don't, and compose run attached, the progress UI will collides with first logs
A fix could be to let the `Full` (tty.go) display be aware we run detached, and not hide events.

**Comment by MaheshThakur9152** [comment]
> this is intentional:
> 
> https://github.com/docker/compose/blob/c8d687599a425502116dac1de9c55d8ecc52bb77/cmd/display/tty.go#L193
> 
> if we don't, and compose run attached, the progress UI will collides with first logs A fix could be to let the `Full` (tty.go) display be aware we run detached, and not hide events.


> Updated! I've reverted the changes in `create.go` and moved the logic to `tty.go` as suggested. The display now allows `StatusStarted` events specifically when running in detached mode.
> Validated with the E2E test. Ready for re-review.

**Review by ndeloof — APPROVED** [review]

File: cmd/compose/compose.go [hunk 3]
```diff
@@ -505,6 +505,7 @@ func RootCommand(dockerCli command.Cli, backendOptions *BackendOptions) *cobra.C
 				display.Mode = display.ModeTTY
 			}
 
+			detached, _ := cmd.Flags().GetBool("detach")
```
Comment (ndeloof) [inline cmd/compose/compose.go:508]: I'm surprised this works, as `detach` flag is declared on child command

File: Makefile [hunk 4]
```diff
@@ -29,6 +29,10 @@ ifeq ($(DETECTED_OS),Windows)
 	BINARY_EXT=.exe
 endif
 
+ifeq ($(DETECTED_OS),Darwin)
```
Comment (ndeloof) [inline Makefile]: this was intentionnaly removed from main branch, now managed within Dockerfile

Comment (MaheshThakur9152) [reply Makefile]: > this was intentionnaly removed from main branch, now managed within Dockerfile

My apologies, that was a remnant from my local dev environment on macOS. I've reverted the Makefile changes. Thanks for catching that

**Review by ndeloof — APPROVED** [review]
